File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Sensors/ExternalEncoder.java
package org.firstinspires.ftc.teamcode.Sensors;

import com.qualcomm.robotcore.hardware.DcMotor;

import org.firstinspires.ftc.teamcode.Control.Constants;

public class ExternalEncoder implements Constants{
    DcMotor motor;
    private int previousPos = 0;
    private long previousTime = 0;
    private double rpm = 0;
    private double power;

    public ExternalEncoder (DcMotor motor) {
        this.motor = motor;
    }

    public int getPosition () {
        return motor.getCurrentPosition();
    }

    public double getRPM() {
        int deltaPos = motor.getCurrentPosition() - previousPos;
        double deltaTime = (System.nanoTime() - previousTime)/NANOSECONDS_PER_MINUTE;
        if (deltaTime*6e4 > 10) {
            rpm = (deltaPos/ E4T_COUNTS_PER_REV)/(deltaTime);
            previousPos = motor.getCurrentPosition();
            previousTime = System.nanoTime();
        }
        return rpm;
    }

    public void reset() {
        motor.setMode(DcMotor.RunMode.STOP_AND_RESET_ENCODER);
        motor.setMode(DcMotor.RunMode.RUN_WITHOUT_ENCODER);
    }

}

File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/Drivers/ExternalEncoder.kt
package org.firstinspires.ftc.teamcode.Drivers

import com.qualcomm.robotcore.hardware.DcMotor

import org.firstinspires.ftc.teamcode.Control.Constants

class ExternalEncoder(internal var motor: DcMotor) : Constants {
    private var previousPos = 0
    private var previousTime: Long = 0
    private var rpm = 0.0
    private val power: Double = 0.0

    val position: Int
        get() = motor.currentPosition

    fun getRPM(): Double {
        val deltaPos = motor.currentPosition - previousPos
        val deltaTime = (System.nanoTime() - previousTime) / Constants.NANOSECONDS_PER_MINUTE
        if (deltaTime * 6e4 > 10) {
            rpm = deltaPos / Constants.E4T_COUNTS_PER_REV / deltaTime
            previousPos = motor.currentPosition
            previousTime = System.nanoTime()
        }
        return rpm
    }

    fun reset() {
        motor.mode = DcMotor.RunMode.STOP_AND_RESET_ENCODER
        motor.mode = DcMotor.RunMode.RUN_WITHOUT_ENCODER
    }

}
